tests/serverpod_test_server/test/entity_relations/table_test.dart

```diff
@@ -0,0 +1,102 @@
+import 'package:serverpod/database.dart';
+import 'package:serverpod_test_server/src/generated/protocol.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('Given entities with deep nested relations', () {
+    group('when building expression with deep nested relations', () {
+      Expression expression =
+          Citizen.t.oldCompany.town.mayor.name.equals('Alex') &
+              Citizen.t.company.name.equals('Serverpod');
+      test('then query prefix is built based on relations.', () {
+        expect(expression.toString(),
+            '(citizen_oldCompany_company_town_town_mayor_citizen."name" = \'Alex\' AND citizen_company_company."name" = \'Serverpod\')');
+      });
+
+      test('then node exists for each sub expression.', () {
+        expect(expression.nodes.length, 4);
+      });
+
+      test('then each column expression has table relation.', () {
+        var columnExpressions = expression.nodes.whereType<Column>();
+        var columnExpressionsWithObjectRelations = columnExpressions
+            .where((element) => element.tableRelations != null);
+
+        expect(columnExpressionsWithObjectRelations.length, 2);
+      });
+    });
+
+    group('when accessing nested relations fields', () {
+      var nestedRelationAccess = Citizen.t.oldCompany.town.mayor.name;
+
+      test('then query prefix is built based on relations.', () {
+        expect(nestedRelationAccess.toString(),
+            'citizen_oldCompany_company_town_town_mayor_citizen."name"');
+      });
+
+      test('then table relations exists.', () {
+        expect(nestedRelationAccess.tableRelations, isNotNull);
+      });
+
+      test('then an table relation is build for each table relation reference.',
+          () {
+        expect(nestedRelationAccess.tableRelations?.length, 3);
```

Comment:
The length asserts that the correct number of table relations are built, we then go on a validate each of the relations so that they contain the correct values in the following tests.

I can add a reason message here but I think this test is required as well.